nilearn/decomposition/_subject_pca.py

```diff
@@ -0,0 +1,330 @@
+        else:
+            try:
+                self.masker_ = clone(self.mask)
+            except TypeError as e:
+                # Workaround for a joblib bug: in joblib 0.6, a Memory object
+                # with cachedir = None cannot be cloned.
+                masker_memory = self.mask.memory
+                if masker_memory.cachedir is None:
+                    self.mask.memory = None
+                    self.masker_ = clone(self.mask)
+                    self.mask.memory = masker_memory
+                    self.masker_.memory = Memory(cachedir=None)
+                else:
+                    # The error was raised for another reason
+                    raise e
+
+            for param_name in ['target_affine', 'target_shape',
+                               'smoothing_fwhm', 'low_pass', 'high_pass',
+                               't_r', 'memory', 'memory_level']:
+                our_param = getattr(self, param_name)
+                if our_param is None:
+                    # Default value
+                    continue
+                if getattr(self.masker_, param_name) is not None:
+                    warnings.warn('Parameter %s of the masker overriden'
+                                  % param_name)
+                setattr(self.masker_, param_name, our_param)
+
+        # Masker warns if it has a mask_img and is passed
+        # imgs to fit().  Avoid the warning by being careful
+        # when calling fit.
+        if self.masker_.mask_img is None:
+            self.masker_.fit(imgs)
+        else:
+            self.masker_.fit()
+        self.mask_img_ = self.masker_.mask_img_
+
+        # Now do the subject-level signal extraction (i.e. data-loading +
+        # PCA)
+        if self.verbose:
+            print("[SinglePCA] Learning subject level PCAs")
+        subject_pcas = Parallel(n_jobs=self.n_jobs, verbose=self.verbose)(
+            delayed(self._cache(session_pca, func_memory_level=1))(
+                img,
+                self.masker_.mask_img_,
+                self._get_filter_and_mask_parameters(),
+                n_components=self.n_components,
+                memory=self.memory,
+                memory_level=self.memory_level,
+                confounds=confound,
+                verbose=self.verbose,
+                random_state=random_state,
+            )
+            for img, confound in zip(imgs, confounds))
+        subject_pcas, subject_svd_vals = zip(*subject_pcas)
+        if not self.whiten:
+            for subject_pca, subject_svd_val in \
+                    zip(subject_pcas, subject_svd_vals):
+                subject_pca *= subject_svd_val[:, np.newaxis]
+
+        self.components_ = list(subject_pcas)
+        self.variance_ = list(subject_svd_vals)
+
+        if len(subject_pcas) == 1:
+            self.components_ = self.components_[0]
+        else:
+            self.components_ = np.empty((len(imgs) * self.n_components,
```

Comment:
> > +            self.components_ = np.empty((len(imgs) * self.n_components,

> Can't you just concatenate the pcas in a one-liner ?

This is actually faster. I believe that it is because numpy has a
sub-optimal memory allocation strategy in the simple concatenation
strategy.